src/p_sparse_matrix.jl

```diff
@@ -1815,6 +1965,182 @@ function psparse_assemble_impl!(B,A,::Type{<:AbstractSplitMatrix},cache)
+        for (i,_,_) in nziterator(A)
+            owner = local_to_owner_row[i]
+            if owner != me
+                ptrs[owner_to_p[owner]+1] += 1
+            end
+        end
+        length_to_ptrs!(ptrs)
+        Tv = eltype(A)
+        ndata = ptrs[end]-1
+        I_snd_data = zeros(Int,ndata)
+        J_snd_data = zeros(Int,ndata)
+        V_snd_data = zeros(Tv,ndata)
+        k_snd_data = zeros(Int32,ndata)
+        for (k,(i,j,v)) in enumerate(nziterator(A))
+            owner = local_to_owner_row[i]
+            if owner != me
+            p = ptrs[owner_to_p[owner]]
+            I_snd_data[p] = local_to_global_row[i]
+            J_snd_data[p] = local_to_global_col[j]
+            V_snd_data[p] = v
+            k_snd_data[p] = k
+            ptrs[owner_to_p[owner]] += 1
+            end
+        end
+        rewind_ptrs!(ptrs)
+        I_snd = JaggedArray(I_snd_data,ptrs)
+        J_snd = JaggedArray(J_snd_data,ptrs)
+        V_snd = JaggedArray(V_snd_data,ptrs)
+        k_snd = JaggedArray(k_snd_data,ptrs)
+        (;I_snd,J_snd,V_snd,k_snd,parts_snd)
+    end
+    function setup_cache_rcv(I_rcv,J_rcv,V_rcv,parts_rcv)
+        k_rcv_data = zeros(Int32,length(I_rcv.data))
+        k_rcv = JaggedArray(k_rcv_data,I_rcv.ptrs)
+        (;I_rcv,J_rcv,V_rcv,k_rcv,parts_rcv)
+    end
+    function finalize_values!(A,rows,cols,cache_snd,cache_rcv)
+        I_rcv_data = cache_rcv.I_rcv.data
+        J_rcv_data = cache_rcv.J_rcv.data
+        V_rcv_data = cache_rcv.V_rcv.data
+        k_rcv_data = cache_rcv.k_rcv.data
+        A_nonzeros = nonzeros(A)
+        map_global_to_local!(I_rcv_data,rows)
+        map_global_to_local!(J_rcv_data,cols)
+        for p in eachindex(k_rcv_data)
+            i = I_rcv_data[p]
+            j = J_rcv_data[p]
+            k = nzindex(A,i,j)
+            @boundscheck @assert k > 0 "The sparsity pattern of the ghost layer is inconsistent"
+            k_rcv_data[p] = k
+            A_nonzeros[k] = f(A_nonzeros[k],V_rcv_data[p])
+        end
+        cache = (;cache_snd...,cache_rcv...)
+        cache
+    end
+    rows = partition(axes(A,1))
+    cols = partition(axes(A,2))
+    parts_snd, parts_rcv = assembly_neighbors(rows)
+    cache_snd = map(setup_cache_snd,partition(A),parts_snd,rows,cols)
+    I_snd = map(i->i.I_snd,cache_snd)
+    J_snd = map(i->i.J_snd,cache_snd)
+    V_snd = map(i->i.V_snd,cache_snd)
+    graph = ExchangeGraph(parts_snd,parts_rcv)
+    t_I = exchange(I_snd,graph)
+    t_J = exchange(J_snd,graph)
+    t_V = exchange(V_snd,graph)
+    @fake_async begin
+        I_rcv = fetch(t_I)
+        J_rcv = fetch(t_J)
+        V_rcv = fetch(t_V)
+        cache_rcv = map(setup_cache_rcv,I_rcv,J_rcv,V_rcv,parts_rcv)
+        cache = map(finalize_values!,partition(A),rows,cols,cache_snd,cache_rcv)
+        if !val_parameter(reuse)
+            A
+        else
+            A, cache
+        end
+    end
+end
+
+# TODO: The Function type parameter can be removed, it's just needed
```

Comment:
Hi @fverdugo , can I get some input on this? 

I've tried to follow the API we currently have for in-place assembly of `PVector`, which takes a binary operator as input that determines how entries get combined into the final array. 
Unfortunately the current assemble API for `PSparseMatrix` has lost this argument, which creates method ambiguities with the methods I have added. 

I can see 3 possible solutions: 

Possible solutions: 
- Remove the binary operator
- Use a different name for the inplace methods
- Also add an operator to the non-inplace methods

Since I've used this feature in the past, I would favor the last two options. Let me know what you think.